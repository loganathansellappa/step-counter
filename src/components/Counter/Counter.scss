.counter {
  display: flex;
  flex-direction: column;
  align-items: center;

  &__step {
    display: flex;
    justify-content: space-evenly;
  }

  &__button {
    background-color: #fff;
    border: 1px solid #ccc;
    border-radius: 3px;
    padding: 5px 50px;
    font-size: 16px;
    cursor: pointer;
    margin-right: 5px;

    &:disabled {
      opacity: 0.5;
      cursor: not-allowed;
    }
  }

  &__input {
    margin: 0 100px;
    padding: 5px 10px;
    border: 1px solid #ccc;
    border-radius: 3px;
    font-size: 16px;
    text-align: center;
  }

  &__error {
    color: red;
    font-size: 14px;
    margin-top: 5px;
  }
  &__value {
    font-size: 20px;
    padding: 10px;
    font-weight: bolder;
    color: #008800;
  }
}
