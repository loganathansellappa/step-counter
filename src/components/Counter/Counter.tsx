import React, { useState } from 'react';
import './Counter.scss';
import { isPositiveNumber } from '~/helpers/utils';

type CounterProps = {
  defaultValue?: number;
};

const Counter: React.FC<CounterProps> = ({ defaultValue = 0 }) => {
  const [counter, setCounter] = useState<number>(0);
  const [step, setStep] = useState<number | string>(() => {
    const savedStep = localStorage.getItem('step');
    return savedStep ? Number(savedStep) : defaultValue;
  });

  const [stepInvalid, setStepInvalid] = useState<boolean>(false);

  const handleIncrement = () => setCounter(counter + Number(step));
  const handleDecrement = () => setCounter(counter - +Number(step));
  const handleReset = () => setCounter(0);

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const inputStep = event.target.value;
    setStep(inputStep);
    if (isPositiveNumber(Number(inputStep))) {
      localStorage.setItem('step', inputStep);
      setStepInvalid(false);
    } else {
      setStepInvalid(true);
    }
  };

  return (
    <div className='counter'>
      <div className='counter__step'>
        <label htmlFor='step-input'>Current Step value</label>
        <input
          className='counter__input'
          type='text'
          id='step-input'
          value={step}
          onChange={handleInputChange}
        />
      </div>
      <div className='counter__buttons'>
        <button className='counter__button' onClick={handleIncrement} disabled={stepInvalid}>
          +
        </button>
        <button className='counter__button' onClick={handleDecrement} disabled={stepInvalid}>
          -
        </button>
        <button className='counter__button' onClick={handleReset}>
          Reset
        </button>
      </div>
      <div className='counter__buttons'>
        <div className='value'>Counter Value: {counter}</div>
      </div>

      {stepInvalid && (
        <div className='error'>Please enter a valid number greater than or equal to 1.</div>
      )}
    </div>
  );
};

export default Counter;
