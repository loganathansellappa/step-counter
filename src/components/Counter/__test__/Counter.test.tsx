import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Counter from '../Counter';
import renderer from 'react-test-renderer';

describe('Counter', () => {
  test('matches the snapshot', () => {
    const tree = renderer.create(<Counter defaultValue={1} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders correctly with default value', () => {
    const { getByText, getByLabelText } = render(<Counter />);
    const incrementButton = getByText('+');
    const decrementButton = getByText('-');
    const resetButton = getByText('Reset');
    const stepInput = getByLabelText('Current Step value');
    const counterValue = getByText('Counter Value: 0');

    expect(incrementButton).toBeInTheDocument();
    expect(decrementButton).toBeInTheDocument();
    expect(resetButton).toBeInTheDocument();
    expect(stepInput).toBeInTheDocument();
    expect(stepInput).toHaveValue('0');
    expect(counterValue).toBeInTheDocument();
  });

  test('increments the counter when the + button is clicked', () => {
    const { getByText, getByLabelText } = render(<Counter />);
    const incrementButton = getByText('+');
    const stepInput = getByLabelText('Current Step value');
    const counterValue = getByText('Counter Value: 0');

    fireEvent.change(stepInput, { target: { value: '2' } });
    fireEvent.click(incrementButton);
    expect(counterValue).toHaveTextContent('2');

    fireEvent.click(incrementButton);
    expect(counterValue).toHaveTextContent('4');
  });

  test('decrements the counter when the - button is clicked', () => {
    const { getByText, getByLabelText } = render(<Counter />);
    const decrementButton = getByText('-');
    const stepInput = getByLabelText('Current Step value');
    const counterValue = getByText('Counter Value: 0');

    fireEvent.change(stepInput, { target: { value: '3' } });
    fireEvent.click(decrementButton);
    expect(counterValue).toHaveTextContent('-3');

    fireEvent.click(decrementButton);
    expect(counterValue).toHaveTextContent('-6');
  });

  test('resets the counter when the Reset button is clicked', () => {
    const { getByText, getByLabelText } = render(<Counter />);
    const incrementButton = getByText('+');
    const resetButton = getByText('Reset');
    const stepInput = getByLabelText('Current Step value');
    const counterValue = getByText('Counter Value: 0');

    fireEvent.change(stepInput, { target: { value: '2' } });
    fireEvent.click(incrementButton);
    fireEvent.click(incrementButton);
    fireEvent.click(resetButton);
    expect(counterValue).toHaveTextContent('0');
  });

  test('updates the step value when the input changes', () => {
    const { getByLabelText } = render(<Counter />);
    const stepInput = getByLabelText('Current Step value');

    fireEvent.change(stepInput, { target: { value: '3' } });
    expect(stepInput).toHaveValue('3');

    fireEvent.change(stepInput, { target: { value: 'abc' } });
    expect(stepInput).toHaveValue('abc');
  });

  test('displays an error message when the step value is invalid', () => {
    const { getByLabelText, getByText } = render(<Counter />);
    const incrementButton = getByText('+');
    const stepInput = getByLabelText('Current Step value');

    fireEvent.change(stepInput, { target: { value: '-1' } });
    fireEvent.click(incrementButton);

    const errorMessage = getByText('Please enter a valid number greater than or equal to 1.');
    expect(errorMessage).toBeInTheDocument();
  });
});
