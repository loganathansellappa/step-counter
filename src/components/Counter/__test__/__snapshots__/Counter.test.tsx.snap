// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter matches the snapshot 1`] = `
<div
  className="counter"
>
  <div
    className="counter__step"
  >
    <label
      htmlFor="step-input"
    >
      Current Step value
    </label>
    <input
      className="counter__input"
      id="step-input"
      onChange={[Function]}
      type="text"
      value={1}
    />
  </div>
  <div
    className="counter__buttons"
  >
    <button
      className="counter__button"
      disabled={false}
      onClick={[Function]}
    >
      +
    </button>
    <button
      className="counter__button"
      disabled={false}
      onClick={[Function]}
    >
      -
    </button>
    <button
      className="counter__button"
      onClick={[Function]}
    >
      Reset
    </button>
  </div>
  <div
    className="counter__buttons"
  >
    <div
      className="value"
    >
      Counter Value: 
      0
    </div>
  </div>
</div>
`;
